crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,314 @@
+  if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL) {
+    return 0;
+  }
+
+  ctx->cipher = EVP_CIPHER_CTX_new();
+  if (ctx->cipher == NULL) {
+    OPENSSL_free(ctx);
+    return 0;
+  }
+  ctx->done = 0;
+  ctx->ok = 1;
+  ctx->buf_off = 0;
+  ctx->buf_len = 0;
+  BIO_set_data(b, ctx);
+  BIO_set_init(b, 1);
+
+  return 1;
+}
+
+static int enc_free(BIO *b) {
+  GUARD_PTR(b);
+
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  EVP_CIPHER_CTX_free(ctx->cipher);
+  OPENSSL_free(ctx);
+  BIO_set_data(b, NULL);
+  BIO_set_init(b, 0);
+
+  return 1;
+}
+
+static int enc_read(BIO *b, char *out, int outl) {
+  GUARD_PTR(b);
+  GUARD_PTR(out);
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL || !ctx->ok) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  int bytes_output = 0;
+  int remaining = outl;
+  const int cipher_block_size = EVP_CIPHER_CTX_block_size(ctx->cipher);
+  while (remaining > 0) {
+    assert(bytes_output + remaining == outl);
+    if (ctx->buf_len > 0) {
+      uint8_t *out_pos = ((uint8_t *)out) + bytes_output;
+      int to_copy = remaining > ctx->buf_len ? ctx->buf_len : remaining;
+      OPENSSL_memcpy(out_pos, &ctx->buf[ctx->buf_off], to_copy);
+      // Update buffer info and counters with number of bytes processed from our
+      // buffer.
+      ctx->buf_len -= to_copy;
+      ctx->buf_off += to_copy;
+      bytes_output += to_copy;
+      remaining -= to_copy;
+      continue;  // continue to see if we've gotten |remaining| to 0
+    }
+    if (!ctx->ok) {
+      ctx->done = 1;  // Set EOF on cipher error.
+      break;
+    }
+    ctx->buf_len = 0;
+    ctx->buf_off = 0;
+    // |EVP_DecryptUpdate| may write up to cipher_block_size-1 more bytes than
+    // requested, so only read bytes we're sure we can decrypt in place.
+    int to_read = (int)sizeof(ctx->buf) - cipher_block_size + 1;
+    int bytes_read = BIO_read(next, ctx->buf, to_read);
+    if (bytes_read == 0 && BIO_eof(next)) {
+      // EVP_CipherFinal_ex may write up to one block to our buffer. If that
+      // happens, continue the loop to process the decrypted block as normal.
+      ctx->ok = EVP_CipherFinal_ex(ctx->cipher, ctx->buf, &ctx->buf_len);
+      // We have more bytes to output, go back to top of loop.
+      if (ctx->buf_len > 0) {
+        continue;
+      }
+      ctx->done = 1;  // If we don't need to output any more bytes, set done.
+      return bytes_output;
+    } else if (bytes_read <= 0) {
+      BIO_copy_next_retry(b);
+      return bytes_output;
+    }
+    // Decrypt ciphertext in place, update |ctx->buf_len| with num bytes
+    // decrypted.
+    ctx->ok = EVP_DecryptUpdate(ctx->cipher, ctx->buf, &ctx->buf_len, ctx->buf,
+                                bytes_read);
+  }
+  return bytes_output;
+}
+
+static int enc_flush(BIO *b, BIO *next, BIO_ENC_CTX *ctx) {
+  GUARD_PTR(b);
+  GUARD_PTR(next);
+  GUARD_PTR(ctx);
+  while (ctx->buf_len > 0) {
```

Comment:
I think this loop instead be conditioned on `ctx->buf_len > 0 || ctx->done == 0`.  It seems possible/likely that `ctx->buf_len == 0` holds when `BIO_flush` is called, in which case this loop never executes (`EVP_CipherFinal_ex` never gets called), and it returns 1 as-if it succeeded.
```
  while (ctx->buf_len > 0 || ctx->done == 0) {
    if (ctx->buf_len > 0) {
...
    } else if(ctx->done == 0) {
...
    }
  }
```

Reply:
i agree, came to this realization as well while reworking enc_flush to work with block ciphers.